support Generate-code-from-TypeSpec doc in js sdk repo (#31949)
Azure/azure-sdk-for-js: documentation/Generate-code-from-TypeSpec.md

# Before you Start

> [TypeScript Azure SDK Design Guidelines](https://azure.github.io/azure-sdk/typescript_introduction.html) is the overall design guideline of the client SDK.

Reviewer: Do we have short url?

Reviewer 2: Agree, suggest to use short link especially for the links to the github repo.

Make sure you are familiar with concepts such as "Service Client" and "Packaging".

Make sure you are familiar with Git.

> See [Prerequisite](https://www.npmjs.com/package/@azure-tools/typespec-ts) of `@azure-tools/typespec-ts`. Make sure that NodeJS, JDK is installed.

Reviewer: Do we need JDK?

# Generate SDK

## Use TypeSpec defined in REST API specifications

:ballot_box_with_check: The package structure in the azure-rest-api-specs repository should follow [these guidelines](https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/typespec-structure-guidelines.md#structure).

> Here is a sample for [tspconfig.yaml](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/contosowidgetmanager/Contoso.Management/tspconfig.yaml).

Reviewer: Can we provide 3 examples, for HLC, RLC and modular client?

Author: hlc doesn't use typespec, I think we can just provide RLC and Modular examples

Please make sure `service-dir`, `package-dir`, `packageDetails`, `isModularLibrary`, `generateMetadata`, `flavor`(for typespec-ts) is correctly configured. `experimentalExtensibleEnums`, `enableOperationGroup`, `hierarchyClient` are the optional configs.
If you want to generate samples with typespec-ts, you should add 
```
examples-directory: "{project-root}/examples"
generateSample:true
```
in your tspconfig.yaml

- "parameters.service-dir.default" would be `sdk/<service>`
- "options.@azure-tools/typespec-ts.package-dir" would be `<module>`

SDK module would be generated under the SDK project folder at `sdk/<service>/<module>`.

> ### Generate Code with code-gen-pipeline tool

Author: wanpeng suggested me to add this way to generated code

> ### Generate Code with code-gen-pipeline tool

Reviewer: It's suggested by Ray to keep consistent in all languages.

Install dependencies to use code-gen-pipeline,  
```ps
npm install -g @azure-tools/typespec-client-generator-cli
npm install -g @microsoft/rush@5.92.0
npm install -g @azure-tools/js-sdk-release-tools
```

> Create a local json file named generatedInput.json with content similar to that shown below

Author: yes this is a required step

# Before you Start

[TypeScript Azure SDK Design Guidelines](https://azure.github.io/azure-sdk/typescript_introduction.html) is the overall design guideline of the client SDK.
Make sure you are familiar with concepts such as "Service Client" and "Packaging".

Make sure you are familiar with Git.

> See [Prerequisite](https://www.npmjs.com/package/@azure-tools/typespec-ts) of `@azure-tools/typespec-ts`. Make sure that NodeJS is installed.

Reviewer 3: What version of NodeJS need to install?

# Generate SDK

## Use TypeSpec defined in REST API specifications

:ballot_box_with_check: The package structure in the azure-rest-api-specs repository should follow [these guidelines](https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/typespec-structure-guidelines.md#structure).

> Here are the samples for\

Reviewer 3: Could you add a section title here for how to configure `tspconfig.yaml`?

[Modular tspconfig.yaml](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/contosowidgetmanager/Contoso.Management/tspconfig.yaml)\
[RLC tspconfig.yaml](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/ai/Face/tspconfig.yaml)

Please make sure `service-dir`, `package-dir`, `packageDetails`, `isModularLibrary`, `generateMetadata`, `flavor`(for typespec-ts) is correctly configured. `experimentalExtensibleEnums`, `enableOperationGroup`, `hierarchyClient` are the optional configs.
If you want to generate samples with typespec-ts, you should add 
```

> examples-directory: "{project-root}/examples"

Author: remove this line

> Please make sure `service-dir`, `package-dir`, `packageDetails`, `isModularLibrary`, `generateMetadata`, `flavor`(for typespec-ts) is correctly configured. `experimentalExtensibleEnums`, `enableOperationGroup`, `hierarchyClient` are the optional configs.

Reviewer 2: Do we have any place to describe the meaning of those configurations?

Reviewer: Looks like Go doc also provide samples only. I think @maririos is working on describing options for all languages, when it's done, we can add the link

generateSample:true
```
in your tspconfig.yaml

- "parameters.service-dir.default" would be `sdk/<service>`
- "options.@azure-tools/typespec-ts.package-dir" would be `<module>`

SDK module would be generated under the SDK project folder at `sdk/<service>/<module>`.

> ### Generate Code with code-gen-pipeline tool

Reviewer: ```
### Generate Code with code-gen-pipeline tool (recommend)
```

> ### Generate Code with code-gen-pipeline tool

Reviewer 3: @kazrael2119 could you refer to the [document for Go](https://github.com/Azure/azure-sdk-for-go/pull/23837/files) to add the corresponding content for JS?

Author: done

Reviewer: @raych1  I think what ziwei authored is better, since we didn't make linux shell be required to customers.

Reviewer 3: Some steps are missing here for JS:
- clone spec repo
- clone JS Sdk repo
- run `sh .scripts/automation_init.sh`

Install dependencies to use code-gen-pipeline,  
```ps

> npm install -g @azure-tools/typespec-client-generator-cli

Reviewer: Please add version `0.14.1`

> **Notice** These step is to generate code with your local spec repo. If you want to generate code with the github url, please refer [Generate Code with tsp-client tool](#Generate-Code-with-tsp-client-tool)

Author: ```suggestion
**Notice** These steps are to generate code using the local spec repo. If you want to generate code with the github url, please refer [Generate Code with tsp-client tool](#Generate-Code-with-tsp-client-tool)
```

npm install -g @microsoft/rush@5.92.0
npm install -g @azure-tools/js-sdk-release-tools
```

Create a local json file named generatedInput.json with content similar to that shown below
```
   {
    "dryRun": false,
    "specFolder": <your-local-spec-repo-path>,
    "headSha": <commit-id-you-want-to-generate-from>,
    "repoHttpsUrl": "https://github.com/Azure/azure-rest-api-specs",
    "relatedTypeSpecProjectFolder": [
    "specification/SERVICE_DIRECTORY_NAME/PACKAGE_DIRECTORY_NAME/"
  ]
}
```

Run the command
```

> code-gen-pipeline --inputJsonPath=<path-to-generatedInput.json> --outputJsonPath={path-to-anywhere(This will not be used locally, but this is required to use code-gen-pipeline)}  --typespecEmitter=@azure-tools/typespec-ts  --local 

Reviewer: Please put it to a quota
```MD

> path-to-generatedOutput.json is the detailed information of generated package, you can ignore it without pipeline.

code-gen-pipeline --inputJsonPath=<path-to-generatedInput.json> --outputJsonPath=<path-to-generatedOutput.json> --typespecEmitter=@azure-tools/typespec-ts --local

```

>     "repoHttpsUrl": "https://github.com/Azure/azure-rest-api-specs",

Reviewer 2: Do we support the private repo or other fork of the spec repo?

Reviewer: This is only used for updating `tsp-location.yaml`. So it can be any repo. 
e.g. https://github.com/Azure/azure-sdk-for-js/blob/b05f9966c17e350548eedeaa6512d98309b8eb34/sdk/openai/openai/tsp-location.yaml#L3
Pleae correct me if I'm wrong. @raych1 .

```

This command will do:\

> 1: generate code with typespec\

Reviewer: step 1: remove exisiting content for this package (if any)

> 1: generate code with typespec\

Reviewer: This command will automatically:
1. generate package code with typespec
1. build package
1. generate tests
1. generate samples if enabled
1. generate CHANGELOG.md
1. bump version based on azure SDK policy 
1. if this is a new package, add this package to project item in rush.json
1. generate/update ci.mgmt.yml or ci.yml (if the generated package is new)

> 2: generate CHANGELOG.md\

Reviewer: + build package as step 2.

3: update rush.json(if the generated package is new)\
4: generate/update ci.mgmt.yml or ci.yml(if the generated package is new)

> ### Generate Code with tsp-client tool

Reviewer: I think we can put it to another doc and call it someting like "advanced way to generate code" since it's only used by a few customer that need to customize their generation

> 

Reviewer: ```
> To reduce workload and unnecessary mistakes, it is recommended to use the simple method from the previous section. Only if you are clear about what you are doing and the method from the previous section does not meet your needs, should you consider using the method below.
```

5. Create or update the `CHANGELOG.md`.
6. Bump the version according to the Azure SDK for JS policy.
7. If this is a new package, add it to the project items in `rush.json`.
8. Generate or update `ci.mgmt.yml` or `ci.yml` (if the package is new).

### Generate Code with tsp-client tool
> To reduce workload and unnecessary mistakes, it is recommended to use the simple method from the previous section. Only if you are clear about what you are doing and the method from the previous section does not meet your needs, should you consider using the method below.

Install `tsp-client` CLI tool

```ps

> npm install -g @azure-tools/typespec-client-generator-cli

Reviewer 2: I heard we shall not use the latest version of tsp-client,  is it also true here?

```
here is a example command of this
```
tsp-client init --tsp-config=../azure-rest-api-specs/specification/connectedcache/ConnectedCache.Management/tspconfig.yaml --local-spec-repo=../azure-rest-api-specs/specification/connectedcache/ConnectedCache.Management --repo=../azure-rest-api-specs --commit=60c67f112c66537f1d006bf1b497857ccd2afacd
```

**Notice**
If you use tsp-client to generate code and your generated SDK is new, you need to do two extra things:

**1**: 
You should add your new SDK in [rush.json](https://github.com/Azure/azure-sdk-for-js/blob/main/rush.json).
Here is the example of the config
```
{
  "packageName": <your-package-name>,
  "projectFolder": <sdk/<service>/<module>>,
  "versionPolicyName": <"management"-or-"client">
}
```

**2**: You should add `ci.yml` or `ci.mgmt.yml` under `sdk/<service>/<module`. `ci.yml` is for `Data Plane SDKs` and `ci.mgmt.yml` is for `Mgmt Plane SDKs`. See [Create/Update the ci.yaml](https://github.com/Azure/azure-sdk-for-js/blob/main/documentation/steps-after-generations.md#createupdate-the-ciyaml)

> # Build

Reviewer: it should be the `####` level section in `### Generate Code with tsp-client tool`, since the `### Generate Code with code-gen-pipeline tool` covers building package.

### How to configure tspconfig.yaml
You can reference these two config files to configure the Modular or RLC package:
- [Modular tspconfig.yaml](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/contosowidgetmanager/Contoso.Management/tspconfig.yaml)
- [RLC tspconfig.yaml](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/ai/Face/tspconfig.yaml)

Please make sure `service-dir`, `package-dir`, `packageDetails`, `isModularLibrary`, `generateMetadata`, `flavor`(for typespec-ts) is correctly configured. `experimentalExtensibleEnums`, `enableOperationGroup`, `hierarchyClient` are the optional configs.
If you want to enable sample generation with typespec-ts, you should add 
```
generateSample:true
```
in your tspconfig.yaml

- "parameters.service-dir.default" would be `sdk/<service>`
- "options.@azure-tools/typespec-ts.package-dir" would be `<module>`

SDK module would be generated under the SDK project folder at `sdk/<service>/<module>`.

### Generate Code with code-gen-pipeline tool (recommend)

> **Notice** These steps are to generate code using the local spec repo. If you want to generate code with the github url, please refer [Generate Code with tsp-client tool](#Generate-Code-with-tsp-client-tool)

Reviewer 4: ```suggestion
**Notice** These steps are to generate code using the local spec repo. If you want to generate code with the github url, please refer [Generate Code with tsp-client tool](#generate-code-with-tsp-client-tool)
```

- [RLC tspconfig.yaml](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/ai/Face/tspconfig.yaml)

Please make sure `service-dir`, `package-dir`, `packageDetails`, `isModularLibrary`, `generateMetadata`, `flavor`(for typespec-ts) is correctly configured. `experimentalExtensibleEnums`, `enableOperationGroup`, `hierarchyClient` are the optional configs.
If you want to enable sample generation with typespec-ts, you should add 
```
generateSample:true
```
in your tspconfig.yaml

- "parameters.service-dir.default" would be `sdk/<service>`
- "options.@azure-tools/typespec-ts.package-dir" would be `<module>`

SDK module would be generated under the SDK project folder at `sdk/<service>/<module>`.

### Generate Code with code-gen-pipeline tool (recommend)

Install dependencies to use code-gen-pipeline,  
```ps
npm install -g @azure-tools/typespec-client-generator-cli@0.14.1
npm install -g @microsoft/rush@5.92.0
npm install -g @azure-tools/js-sdk-release-tools
```

Create a local json file named generatedInput.json with content similar to that shown below
```

>    {

Reviewer 4: wonder if we could format this code block ?
